### ISSUE TITLE
Semigroupoids

### ISSUE THREAD

**Issue by freebroccolo** [freebroccolo/agda-groupoids#35]
I think the `(n, i, r)`-category approach I described earlier might turn out to be fairly usable. There are two problems though: both the inverse coherence constraint cells and isomorphisms would need to be reworked to not rely on identities. This should be possible by restating the laws in terms of morphisms that cancel on the left or right of composition or the fact that `f ⟔ f⁻¹ ⟔ f ≃ f` which amounts to the same thing. It should be possible in the case where it is known that `i = 0` to derive these from the standard versions or vice versa.

/cc @jonsterling

**Comment by jonsterling** [comment]
OK, this sounds very promising then.

**Comment by freebroccolo** [comment]
I am thinking of something like the following. It's a little unfortunate that two additional cells are required for the inverses but I don't see an easy way around that. @jonsterling do you know if something similar is going on for other higher dimensional PER stuff? (to the extent that it has been worked out…)

```agda
record 𝔘 (i↑ r↑ : ℕ∞) ..ℓ : Set (lsuc ℓ) where
  no-eta-equality
  infix 3 [_]
  open Fin
  field
    [_] : Gph ∞ i↑ r↑ ℓ

  complex = [_]

  open Gph public
  open Cell {𝒢 = complex}

  field
    idn₀
      : {≜i : i↑ T.≡ 0}
      → {a : ⊢*}
      → 1 ⊢ a ↝ a
    seq₀
      : {a b c : ⊢*}
      → (f : 1 ⊢ a ↝ b)
      → (g : 1 ⊢ b ↝ c)
      → 1 ⊢ a ↝ c
    inv₀
      : {≜r : r↑ T.≡ 0}
      → {a b : ⊢*}
      → (f : 1 ⊢ a ↝ b)
      → 1 ⊢ b ↝ a
  field
    seq₀*
      : {a b c : ⊢*}
      → {f₀ f₁ : 1 ⊢ a ↝ b}
      → {g₀ g₁ : 1 ⊢ b ↝ c}
      → (α : 2 ⊢ f₀ ↝ f₁)
      → (β : 2 ⊢ g₀ ↝ g₁)
      → 2 ⊢ seq₀ f₀ g₀ ↝ seq₀ f₁ g₁
    inv₀*
      : {≜r : r↑ T.≡ 0}
      → {a b : ⊢*}
      → {f g : 1 ⊢ a ↝ b}
      → (α : 2 ⊢ f ↝ g)
      → 2 ⊢ inv₀ {≜r} f ↝ inv₀ {≜r} g
  field
    ⊢idn₀-λ
      : {≜i : i↑ T.≡ 0}
      → {a b : ⊢*}
      → {f : 1 ⊢ a ↝ b}
      → 2 ⊢ seq₀ (idn₀ {≜i}) f ↝ f
    ⊢idn₀-ρ
      : {≜i : i↑ T.≡ 0}
      → {a b : ⊢*}
      → {f : 1 ⊢ a ↝ b}
      → 2 ⊢ seq₀ f (idn₀ {≜i}) ↝ f
    ⊢seq₀-α
      : {a b c d : ⊢*}
      → {f : 1 ⊢ a ↝ b}
      → {g : 1 ⊢ b ↝ c}
      → {h : 1 ⊢ c ↝ d}
      → 2 ⊢ seq₀ f (seq₀ g h) ↝ seq₀ (seq₀ f g) h
    ⊢inv₀-λ-λ
      : {≜r : r↑ T.≡ 0}
      → {a b c : ⊢*}
      → {f : 1 ⊢ a ↝ b}
      → {k : 1 ⊢ b ↝ c}
      → 2 ⊢ seq₀ (seq₀ (inv₀ {≜r} f) f) k ↝ k
    ⊢inv₀-λ-ρ
      : {≜r : r↑ T.≡ 0}
      → {a b c : ⊢*}
      → {f : 1 ⊢ a ↝ b}
      → {k : 1 ⊢ a ↝ c}
      → 2 ⊢ seq₀ (seq₀ f (inv₀ {≜r} f)) k ↝ k
    ⊢inv₀-ρ-λ
      : {≜r : r↑ T.≡ 0}
      → {a b c : ⊢*}
      → {f : 1 ⊢ a ↝ b}
      → {k : 1 ⊢ c ↝ b}
      → 2 ⊢ seq₀ k (seq₀ (inv₀ {≜r} f) f) ↝ k
    ⊢inv₀-ρ-ρ
      : {≜r : r↑ T.≡ 0}
      → {a b c : ⊢*}
      → {f : 1 ⊢ a ↝ b}
      → {k : 1 ⊢ c ↝ a}
      → 2 ⊢ seq₀ k (seq₀ f (inv₀ {≜r} f)) ↝ k
  field
    idn₁
      : {a b : ⊢*}
      → {f : 1 ⊢ a ↝ b}
      → 2 ⊢ f ↝ f
    seq₁
      : {a b : ⊢*}
      → {f g h : 1 ⊢ a ↝ b}
      → (α : 2 ⊢ f ↝ g)
      → (β : 2 ⊢ g ↝ h)
      → 2 ⊢ f ↝ h
    inv₁
      : {a b : ⊢*}
      → {f g : 1 ⊢ a ↝ b}
      → (α : 2 ⊢ f ↝ g)
      → 2 ⊢ g ↝ f
```

**Comment by jonsterling** [comment]
Hm, I'll have to think about it a bit. Re cubical/higher PERs, I don't really know—I doubt that much attention is being paid at the moment to this side of the apparatus, tbh.

By the way, I was a little bit confused about how this development fits into standard definitions of `(n,r)`-categories that I've seen. For instance, in the defn. I'm familiar with, something is an `(n,r)` category if all morphisms above level `n` are trivial and all morphisms above level `r` are invertible; but here, you're just checking that `r` is `0` before requiring invertibility. Is this equivalent in some sense? Or am I looking at it from the wrong end?

**Comment by freebroccolo** [comment]
> By the way, I was a little bit confused about how this development fits into standard definitions of (n,r)-categories that I've seen. For instance, in the defn. I'm familiar with, something is an (n,r) category if all morphisms above level n are trivial and all morphisms above level r are invertible; but here, you're just checking that r is 0 before requiring invertibility. Is this equivalent in some sense? Or am I looking at it from the wrong end?

Yes, that's the right idea and I think this version should be equivalent. The basic idea is that you are always looking at one particular set of edges of the hom complex at some point in the sequence. The `n, i, r` parameters act like the "distance" to the edges in the sequence after which the operations are always defined. The parameters are decreased each time you move to a higher dimension, which you can see [here](https://github.com/freebroccolo/agda-globular-objects/blob/32981c8f46cf7a3a0b2fbb5f5343856ebc8c4acd/src/Globular/Quiver/Ordinary/Universe/Boot.agda#L17). The downward arrow is conatural predecessor, which is the identity on `0` and `∞`.

So you can read these `{≜r : r↑ ℕ∞.≲ 0}` arguments as being proofs that the operation is defined at the particular stage you are calling them at.

For the current implementation, we don't make much use of the fact that we can work with all of `ℕ∞`, but it's not really any more complex than working with some small finite value.

**Comment by freebroccolo** [comment]
I should note that the predecessor is identity on `0` and `∞` up to bisimulation.

**Comment by jonsterling** [comment]
Ah, this clarifies a lot for me, thanks.

**Comment by freebroccolo** [comment]
By the way, after experimenting with it some more, I think I'm going to set aside the idea of implementing things in the current library this way. I do think it would clarify a number of things but it's a bit more complexity than I want to introduce at this stage.

The main advantage to having done it this way would be a (mostly) uniform set of constructions and operations across setoids, PERs, categories, groupoids, and some other structures in that space. But I'm not sure how useful PERs or semigroupoids would be yet for general constructions aside from stuff specifically related to type theory and semantics. So I think it might be a better idea to keep things simple and just implement PERs in a separate library and use this library to reason about them in the standard categorical fashion (well, as standard as the current approach is anyway).

We could still implement them on top of the globular quivers now that I've made them non-reflexive and it would already provide a lot of the basic stuff (more or less we already have a cosmos of quivers). Essentially they'd be like the structure I've described above but without any of the higher constraint cells and without identity.

Once the category of PERs in this fashion is defined, I suppose we'd have presheaf style cubical PERs.
